### ISSUE TITLE
Worker tries to execute the workflow even if pulling image from registry fails

### ISSUE THREAD

**Issue by invidian** [tinkerbell/tink#382]
## Expected Behaviour

Workflow execution should stop if image can be pulled.

## Current Behaviour

Even if pulling image from registry fails, worker still tries to execute the workflow and then complains that the image cannot be found.

## Possible Solution

Make sure pulling errors are handled properly.

## Steps to Reproduce (for bugs)

1. Create a worker machine with 4GB of RAM to make `tmpfs` in OSIE of size 2GB.
2. Create a workflow with Docker image bigger than 1GB (e.g. `ubuntu` with `qemu` and `libguestfs-tools` packages installed).
3. Try running a workflow.

**Comment by Cbkhare** [comment]
hey @invidian ,  nice catch! 

I tried reproducing the issue and below are my findings. 

- The error which is visible to us is actually not the root cause. Workflow execution has failed way before while pulling the image. It can be seen in the tink-worker logs with message `no space left on the device`. 
- What we are observing is that the exit should have been before creating the container, which didn't really happened. 
- Reason for the problem here is that [io.copy](https://github.com/tinkerbell/tink/blob/a41538da8b478b4d46aa5ab88d7dccfb54f9273d/cmd/tink-worker/internal/registry.go#L67) never really received the syscall of failure and eventually didn't returned the same. 
- And lately, our workflow execution went ahead for [creating](https://github.com/tinkerbell/tink/blob/master/cmd/tink-worker/internal/worker.go#L104) the container and failed. 


I tried to pinpoint the error by debugging it with delve. What seems to cause the failure is during the [Copy](https://github.com/golang/go/blob/master/src/io/io.go#L429). We can notice in the below logs after docker interrupt, Read returned `EOF` which for copy is graceful exit instead of an error. Usually, io.Copy returns the error when failed for `no space left on the device` but with docker pull I am expecting the interrupt is not received by the go compiler. 


```{"status":"Extracting","progressDetail":{"current":375681641,"total":375681641},"progress":"[==================================================\u003e]  375.7MB/375.7MB","id":"7899dcd9f509"}
{"errorDetail":{"message":"failed to register layer: Error processing tar file(exit status 1): write /root/test123/cdb5015ff36390a47e559aafbbc8f0713264b8815717f21c5db8026a0656f818/layer.tar: no space left on device"},"error":"failed to register layer: Error processing tar file(exit status 1): write /root/test123/cdb5015ff36390a47e559aafbbc8f0713264b8815717f21c5db8026a0656f818/layer.tar: no space left on device"}
> os.(*File).Write() /usr/local/go/src/os/file.go:175 (PC: 0x51bb0b)
   170:	func (f *File) Write(b []byte) (n int, err error) {
   171:		if err := f.checkValid("write"); err != nil {
   172:			return 0, err
   173:		}
   174:		n, e := f.write(b)
=> 175:		if n < 0 {
   176:			n = 0
   177:		}
   178:		if n != len(b) {
   179:			err = io.ErrShortWrite
   180:		}
(dlv) print n
990
(dlv) print e
error nil
(dlv) print len(b) 
990
> os.(*File).Write() /usr/local/go/src/os/file.go:188 (PC: 0x51bbf4)
   183:	
   184:		if e != nil {
   185:			err = f.wrapErr("write", e)
   186:		}
   187:	
=> 188:		return n, err
   189:	}
   190:	
   191:	var errWriteAtInAppendMode = errors.New("os: invalid use of WriteAt on file opened with O_APPEND")
   192:	
   193:	// WriteAt writes len(b) bytes to the File starting at byte offset off.
(dlv) n
> io.copyBuffer() /usr/local/go/src/io/io.go:425 (PC: 0x504ec2)
Values returned:

   420:			buf = make([]byte, size)
   421:		}
   422:		for {
   423:			nr, er := src.Read(buf)
   424:			if nr > 0 {
=> 425:				nw, ew := dst.Write(buf[0:nr])
   426:				if nw < 0 || nr < nw {
   427:					nw = 0
   428:					if ew == nil {
   429:						ew = errInvalidWrite
   430:					}
(dlv) n
(dlv) print nw
Command failed: could not find symbol value for nw
(dlv) print nr
990
(dlv) print er
error(*errors.errorString) *{s: "EOF"}
(dlv) n
> io.copyBuffer() /usr/local/go/src/io/io.go:447 (PC: 0x505032)
   442:			}
   443:			if er != nil {
   444:				if er != EOF {
   445:					err = er
   446:				}
=> 447:				break
   448:			}
   449:		}
   450:		return written, err
   451:	}
   452:	
(dlv) print er
error(*errors.errorString) *{s: "EOF"}
(dlv) print err
error nil
```

**Comment by Cbkhare** [comment]
resolved after #450 merge.
